pandas/core/frame.py

```diff
@@ -3485,8 +3499,30 @@ def unstack(self, level=-1):
         -------
         unstacked : DataFrame or Series
         """
-        from pandas.core.reshape import unstack
-        return unstack(self, level)
+        #from pandas.core.reshape import unstack
+        #return unstack(self, level)
+        result = self.T.stack(level, dropna=False, sequentially=sequentially).T
+
+        if isinstance(result, DataFrame):
```

Comment:
Well, the idea is to try to restore the original ``dtypes`` (e.g. an integer or boolean type) unless can't do so due to the introduction of ``NaN``s. This was not the case previously -- see, for example, my change to ``pandas/tools/tests/test_pivot.py`` below. I'd welcome a cleaner way of doing this.